### ISSUE TITLE
Question about `OpCodes.Unaligned` prefix.

### ISSUE THREAD

**Issue by sandreenko** [dotnet/roslyn#43558]
**Version Used**: 
5.0.100-preview.4.20202.8
**Steps to Reproduce**:

Compile a test like this:
```
        const int arrSize = 100;
        const int floatSize = 4;
        const int doubleSize = 8;

        byte[] arr = new byte[arrSize];

        for (int i = 0; i < arrSize - floatSize; ++i)
        {
            ref byte r = ref Unsafe.Add<byte>(ref arr[0], i);
            var p = Unsafe.AsPointer(ref r);
            float* f = (float*)p;
            *f = 1;
            Debug.Assert(*f == 1);
        }
```
For `*f = 1` you will see IL like:
```
    [ 0]  36 (0x024) ldloc.s 4
    [ 1]  38 (0x026) ldc.r4 1.0000000000000000
    [ 2]  43 (0x02b) stind.r4
```
**Expected Behavior**:
I have expected `OpCodes.Unaligned` prefix for that `stind.r4` because `ldloc.s 4` address is not aligned to the float natural size.

**Actual Behavior**:
There is not `Unaligned` prefix, so on arm32 Jit generates load and stores that work only with aligned address and fail at runtime:
```
System.DataMisalignedException: A datatype misalignment was detected in a load or store instruction.
   at FloatNonAlignedFieldWithSmallOffset..ctor(Single a)
   at Test.Main()
```

Does this prefix sound like something that Roslyn should set in this case?

**Comment by GrabYourPitchforks** [comment]
Are you suggesting that the compiler should always pessimistically emit an `unaligned.` prefix for any `ldind` / `stind` instruction? Normally such references can only be generated via unsafe code, but I guess `[FieldOffset]` is also a thing that people might use. :)

**Comment by sandreenko** [comment]
I am not suggesting, I am just asking how that code should work according to the specs. As I see from [ECMA ](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf) III.2.5 unaligned. (prefix) it should be set in this case, am I reading it correctly?

Note: jit will ignore that prefix for all platforms except ARM32 (for floating-point stores/loads) and ARM64 ( for accesses that are also marked as VOLATILE).

**Comment by tannergooding** [comment]
I believe that is the purpose of `Unsafe.ReadUnaligned` and `Unsafe.WriteUnaligned` and users who are purposefully misaligning data or who might be getting unaligned data should use them to account for the difference.

You would likewise hit it with `FieldOffset(1)` or any number of other patterns in "safe" code.

**Comment by sandreenko** [comment]
> but I guess [FieldOffset] is also a thing that people might use. :)

I was trying to do that without `Unsafe` using `[FieldOffset]` but did not find a way to break it.
```
struct SimpleStruct
{
    // the field is aligned inside SimpleStruct.
    public float field;
}
[StructLayout(LayoutKind.Explicit)]
internal struct StructNonAlignedField
{
    // SimpleStruct is unaligned, so the result offset to its float field is unaligned.
    [FieldOffset(1)]
    public SimpleStruct field;
    public StructNonAlignedField(float a)
    {
        field.field = a;
    }
}
```
for that case IL access is using `StructNonAlignedField` address as its base, so the final offset is misaligned, but `an address currently atop the evaluation stack` is address of the parent struct, it is aligned, so the prefix is not required, JIT is finding that itself and generate correct access.

**Comment by sandreenko** [comment]
so maybe I can rephrase my question as:
1. Is it something that Roslyn can fix?
2. Is it something that Jit should fix (in that case it won't be clear for me why we need that prefix at all)?
3. Is it like C++ with undefined behavior and nobody needs to do anything?

**Comment by GrabYourPitchforks** [comment]
Consider the following sample. What does it do on an ARM32 box where misaligned accesses are forbidden by default? (I don't have equipment to test this.)

```cs
[StructLayout(LayoutKind.Explicit)]
struct MyStruct
{
    [FieldOffset(1)]
    public float _f;
}

public static void Main()
{
    MyStruct s = new MyStruct();
    Console.WriteLine(Deref(ref s._f));
}

[MethodImpl(MethodImpOptions.NoInlining)]
private static float Deref(ref float)
{
    return float;
}
```

**Comment by sandreenko** [comment]
> Consider the following sample. What does it do on an ARM32 box where misaligned accesses are forbidden by default? (I don't have equipment to test this.)

If you want you can use `set Complus_altjit=*; set Complus_altjitname=protononjit.dll; set complus_jitdump=Main;` to see what Jit generates for this case (running x86 Runtime repo).

For that case, Jit will generate load/store from float to int register and then generate int value store/load (INS_str or INS_ldr, they support misaligned addresses), the logic is here https://github.com/dotnet/runtime/blob/90cc6dff11e4eea07d8b2ed438b25679f0caf7a8/src/coreclr/src/jit/codegenarm.cpp#L1006.

Jit considers an address that comes from IL without `Unaligned` prefix as aligned and tracks offsets that the program adds to the address. If the sum offset is not aligned than we generate unaligned access, in your example offset would be `1`.

The situation in the issue example is different, IL gives us an address that is already unaligned, but doesn't have the prefix.

**Comment by GrabYourPitchforks** [comment]
The `SetVal` method doesn't know whether the incoming _ref_ is misaligned or not, though. It's just a single `stind.r4` instruction with no context as to where the target lives. So for this to work in all cases, wouldn't the JIT have to pessimistically assume when producing the codegen for `SetVal` that _ref val_ might be misaligned, so it'll always have to bounce through the integer registers?

**Comment by sandreenko** [comment]
You are right, jit won't catch it if `stdind.r4` is not marked in IL:
```
IN0004: 000000  4B      push    {r11,lr}
IN0005: 000004  2B      mov     r11, sp

G_M38788_IG02:        ; offs=000006H, size=000CH, bbWeight=1    PerfScore 3.00, gcrefRegs=0000 {}, byrefRegs=0001 {r0}, byref

IN0001: 000006  4B      mov     r3, 0x3f800000
IN0002: 00000A  4B      vmov.i2f s8, r3
IN0003: 00000E  4B      vstr    s8, [r0] <- misaligned runtime exception

G_M38788_IG03:        ; offs=000012H, size=0004H, bbWeight=1    PerfScore 1.00, epilog, nogc, extend

IN0006: 000012  4B      pop     {r11,pc}
```

**Comment by BruceForstall** [comment]
I don't think `SetVal` requires an `unaligned` prefix for the `stind` -- but the user either needs to do something special to force one, or expect getting an alignment fault.

**Comment by sandreenko** [comment]
> I don't think SetVal requires an unaligned prefix

Does not it contradict to this ECMA statement:
> Code generators that do not restrict their output to a 32-bit word size (see Partition I and Partition II) shall use `unaligned.` if the alignment is not known at compile time to be 8-byte. 

?

I think the compiler in this context is Roslyn.

**Comment by BruceForstall** [comment]
Perhaps Roslyn should emit a warning/error when it compiles the `SetVal(ref s._f);` call because it knows then that it's passing an unaligned address to a function that expects an aligned address.
